### PR TITLE
Fix last_harvest.txt datestamp incorrectly updated when OAI-PMH response is not chronological

### PR DESCRIPTION
**Pull request by ryan-jacobs**
Addresses issue #7 

Sets last_harvest.txt datestamp to an explicit date and time based on the harvest range specified at harvest launch instead of the individual datestamps returned during harvest (which may not be chronological).

TODO
- [ ] Update changelog
- [ ] Release new major version after merging

### CODE REVIEW

**Comment by demiankatz** [comment]
Thanks for this, @ryan-jacobs. I've left some more detailed comments on #7. I've also allowed the continuous integration tasks to run, which reveals (unsurprisingly) that these changes break some tests. Once we settle on an approach, we'll also need to update the test suite before we can merge this. Please let me know if you need any help with that!

File: src/OaiPmh/Harvester.php
Comment (demiankatz) [inline src/OaiPmh/Harvester.php]: I think you can safely remove this method rather than just deprecating it. We'll need to release this as a major version due to backward-breaking changes, and I don't think there's any value in retaining the legacy function.

File: src/OaiPmh/Harvester.php
Comment (demiankatz) [inline src/OaiPmh/Harvester.php]: As discussed in #7, I still think we're better off leaving things open-ended when no until is specified. What you're doing here makes perfect sense within the specification, but there are a lot of faulty implementations out there, so it's better to err on "too many records" than "not enough records."

File: src/OaiPmh/Harvester.php
Comment (demiankatz) [inline src/OaiPmh/Harvester.php]: For backward compatibility, I wonder if we should account for the possibility of $this->harvestEndDate being undefined here (in case a user upgrades and then resumes an interrupted harvest from a prior version). This is probably a pretty rare case, and maybe the thing to do is just throw an exception saying the harvest needs to be restarted.. but it might be worth accounting for the possibility in some way.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks for the ongoing work, @ryan-jacobs, I think this is shaping up very nicely, and I definitely agree with your decision to consolidate logic into a single date-related method now that I see what you meant. Just a few comments related to the ongoing conversation...

File: src/OaiPmh/Harvester.php
Comment (ryan-jacobs) [reply src/OaiPmh/Harvester.php]: Great. This was my instinct as well, especially as I don't see any way for folks to dynamically extend this Harvester class in their own implementations.

Thinking further, while I've preserved (moved) the logic to set the granularity property, it does not look like the property itself is actually used anymore. It used to be part of the final date calculation for last_harvest.txt (in [getRecords()](https://github.com/vufind-org/vufindharvest/blob/dev/src/OaiPmh/Harvester.php#L271)), but that's getting replaced as a fundamental part of this PR. I think the whole property could be dropped. Can you see any forward-compatible justification to keep it?

File: src/OaiPmh/Harvester.php
Comment (ryan-jacobs) [reply src/OaiPmh/Harvester.php]: While I also suspect that's a pretty rare case, it certainly could not hurt. Perhaps checking to see that we have a state array with 4 values will be most useful, as that would catch any strange situations where an incomplete state was saved for any reason.

File: src/OaiPmh/Harvester.php
Comment (ryan-jacobs) [inline src/OaiPmh/Harvester.php]: Public note to self... I'm actually thinking that this call should be moved inside of the launch() method. storeDateSettings(), and previously loadgranularity() (where this call originally lived), get called from the constructor. This feels like something that should be moved outside of the constructor, especially as an external call is needed. Having it inside launch() may also make managing returned values a bit easier, since that's the scope where they will most likely be needed.

**Review by ryan-jacobs — COMMENTED** [review]
Thanks @demiankatz for your comments. I'm hoping to pick this up again next week. Some other work commitments got in the way this week. I've added some responses here along with a couple self-review notes.

File: src/OaiPmh/Harvester.php
Comment (ryan-jacobs) [reply src/OaiPmh/Harvester.php]: I agree, and I appreciate the discussion that led to this conclusion.

File: src/OaiPmh/Harvester.php
Comment (demiankatz) [reply src/OaiPmh/Harvester.php]: In theory, it is possible to dynamically extend the Harvester in VuFind (you can override the configuration for the harvest command and replace it with your own subclass and factory), but in practice it would be a lot of work since you'd have to override a lot of pieces to make it work. This perhaps highlights the fact that our factory could be made more flexible by replacing some hard-coded class names with configurable properties... but that's another conversation for another thread. :-)

We might need to think a little more about the granularity setting. That is currently a setting that gets loaded from oai.ini or console parameters (see https://github.com/vufind-org/vufindharvest/blob/dev/src/OaiPmh/HarvesterCommand.php#L226-L230). The original purpose of this was that if you knew and specified the granularity in advance, we could bypass the Identify call. Now that we have a more frequent reason to call Identify, the value of that is significantly reduced, but I think there are potentially still situations where it applies -- i.e. if the granularity is specified in advance and we're given an "Until" date, I think it's still possible to proceed without calling Identify.

File: src/OaiPmh/Harvester.php
Comment (demiankatz) [reply src/OaiPmh/Harvester.php]: Works for me!

File: src/OaiPmh/Harvester.php
Comment (demiankatz) [reply src/OaiPmh/Harvester.php]: Agreed, it's fairly bad form for the constructor to have side effects. It might also make sense to set up smart getters -- i.e. rather than fetching the Identify response in advance, we have a getIdentifyResponse() method that returns a property if defined or pulls from the server if undefined. Then anything that needs data from there can call getIdentifyResponse... and then if the code gets reorganized or refactored, it will still be smart enough to pull what it needs on demand and not do redundant work.

File: src/OaiPmh/Harvester.php
Comment (ryan-jacobs) [reply src/OaiPmh/Harvester.php]: https://github.com/vufind-org/vufindharvest/pull/8/commits/cb0d8cc360010573e1a8e1d47c92c3469bfe14ab addresses this piece.

**Comment by ryan-jacobs** [comment]
I just wanted to note that I'm still committed to addressing the changes discussed, just for sidetracked with other work and personal commitments.

**Comment by demiankatz** [comment]
No worries, @ryan-jacobs, I am very familiar with that situation. ;-) I appreciate the help and will be patient awaiting its completion!

File: src/OaiPmh/Harvester.php
Comment (ryan-jacobs) [reply src/OaiPmh/Harvester.php]: Yes, I think that will improve the DX quite a bit. https://github.com/vufind-org/vufindharvest/pull/8/commits/569569a47dd72eae273aec256a15cae07c093a9e adds a getIdentifyResponse() method such that we can get related properties (over and over) with nice chained calls, like:
```
$this->getIdentifyResponse()->granularity;
$this->getIdentifyResponse()->responseDate;
```

File: src/OaiPmh/Harvester.php
Comment (ryan-jacobs) [reply src/OaiPmh/Harvester.php]: Yes, this is interesting. I'm still trying to untangle exactly what we need to be doing with that granularity value (no matter how it's sourced). Previously we were using it to convert a non-UTC timestamp from record headers into a UTCdatetime format that can be used in last_harvest.txt. Now however, the last_harvest.txt value is based on a server datestamp that's already UTC, or (seemingly very rarely) a direct user-supplied "until" value. I'm wondering if extra care and validation is actually needed to convert all selective harvesting values into the right granularity. The OAI spec just seems to indicate that "from" and "until" values need to be UTCdatetime format, but I don't see any requirement about exactly matching the server's granularity setting for those. I thought this granularity was just an indication of what behavior to expect, as opposed to a strict input format requirement.

As it stands right now in the PR the "until" value is based on last_harvest.txt, which is now tied to the previous harvest's server identify responseDate value. This is always UTC in _second_ granularity. I wonder if later passing this as a "from" value to an OAI host with "day" granularity will cause issues. If so, we probably need some sort of date validate/convert method to check the format against the granularity. I suppose that any user-supplied "from" and "until" values could also be passed though this.

File: src/OaiPmh/Harvester.php
Comment (ryan-jacobs) [reply src/OaiPmh/Harvester.php]: https://github.com/vufind-org/vufindharvest/pull/8/commits/ccd0f59ff3581c302a917b2b4a1e93079b3fb990 now incorporates some changes directly to launch() that preserves the old null handling for "until". When this value is null, last_harvest.txt is sourced separately from the OAI server's clock. All the state preservation logic is still preserved. I think this would give us the best of all worlds as per the issue discussion.

**Comment by ryan-jacobs** [comment]
@demiankatz I finally got some time today to get back into the core elements of this. I think all comments are either resolved, or acknowledged with new questions (like the granularity topic). Let me know how things look from your end. I'll defer jumping into test fixes until most implementation specifics are more settled.

File: src/OaiPmh/Harvester.php
Comment (demiankatz) [inline src/OaiPmh/Harvester.php]: Typo: relaibility

File: src/OaiPmh/Harvester.php
Comment (demiankatz) [inline src/OaiPmh/Harvester.php]: There are some minor style things (we use lowercase rather than uppercase booleans, for example) that you can fix with `vendor/bin/phing php-cs-fixer` -- that will help ensure that the CI build passes. You might also want to try `vendor/bin/phpcs-console` to see if there are any other style issues (like very long lines) that require manual fixes (or, of course, you could just look at the build output on this PR if that's easier).

File: src/OaiPmh/Harvester.php
Comment (demiankatz) [reply src/OaiPmh/Harvester.php]: Ahh, you're right, it does appear that maybe the whole granularity thing can simply be removed at this point. It looks to me like the purpose of it was to ensure that dates we generated were consistent with dates received from the server... but since all that logic is gone, it no longer matters.

File: src/OaiPmh/Harvester.php
Comment (demiankatz) [reply src/OaiPmh/Harvester.php]: Yes, makes sense to me!

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks for the progress on this; I think it's nearly done apart from style/test tweaks, and possibly removing obsolete granularity logic (though I'm also open to more fully addressing that in a follow-up PR if you think it would be cleaner that way).

I think I've provided enough guidance to move things forward, but if there's more I can do to help, just let me know!

File: src/OaiPmh/Harvester.php
Comment (ryan-jacobs) [reply src/OaiPmh/Harvester.php]: Addressed in https://github.com/vufind-org/vufindharvest/pull/8/commits/c40d5748c210af0cf96647d28a7d70e17a839c10

File: src/OaiPmh/Harvester.php
Comment (ryan-jacobs) [reply src/OaiPmh/Harvester.php]: Ah yes, thanks for the php-cs-fixer tip. I also see a couple places with incorrect indents. https://github.com/vufind-org/vufindharvest/pull/8/commits/c40d5748c210af0cf96647d28a7d70e17a839c10 should have taken care of all these.

File: src/OaiPmh/Harvester.php
Comment (ryan-jacobs) [reply src/OaiPmh/Harvester.php]: Thanks for checking my reasoning here! While this may contradict my “let the spec be the main guide” tone from earlier comments, I’m still mindful of that one case where an OAI-PMH host with day granularity starts getting a “from” value with second granularity. While the spec indicates that this certainly should not matter, one takeaway from these conversations is that OAI-PMH hosts may behave inconsistently. Reflecting more -- if there is an edge case where this actually causes a problem, we could be introducing a regression.

While enforcing UCT the previous implementation did also flip a system-provided date into YYYY-MM-DD format before writing the value to last_harvest.txt when day granularity was detected. We could (should?) continue to do that with little risk. I had considered some sort of date format checker/filter for all values sent as “from” and “until” params, but given the way direct user inputs could be used for those (with varying formats and timezone encodings), that seems like a risky proposition. The new Identify responseDate format is much more predictable though, and so a granularity check/conversion on it seems doable within a context likes launch(), where we know `$explicitHarvestEndDate` is only being used to source last_harvest.txt data:

```
if (empty($this->harvestEndDate)) {
    $explicitHarvestEndDate = $this->getIdentifyResponse()->responseDate;
    // Backwards compatible support for any OAI-PMH hosts that may require
    // day granularity format for all date params.
    $granularity = $this->granularity == 'auto' ? $this->getIdentifyResponse()->granularity : $this->granularity;
    if ($granularity == 'YYYY-MM-DD') {
      $explicitHarvestEndDate = (new \DateTime($explicitHarvestEndDate))->format('Y-m-d');
    }
} else {
    $explicitHarvestEndDate = $this->harvestEndDate;
}
```
Is that worth considering as an addition? I don't want to over-analyze an edge case here, but also want to be respectful of these nuances from the previous implementation if needed.

Comment (demiankatz) [reply src/OaiPmh/Harvester.php]: Good point. My original thought was that if we're feeding back data the system provides to us, it should be able to understand its own output as input, but you're right that we can't really make this assumption -- the system could potentially be able to report its own clock time in full granularity while serving up data with lesser granularity... and the spec says that an OAI server MUST throw an error if it receives from/until parameters in the wrong granularity. Thus, I believe you are correct that we still need the granularity check. Your code above makes sense to me, though given the predictability of the formats, maybe it would be better to trim the UTCdate with a substr() rather than doing a full date conversion. That would reduce chances of something unpredictable happening and also require less processing. (Obviously it could cause trouble if the server returns malformed date values, but I think that's going to cause trouble in any situation).

**Comment by demiankatz** [comment]
I see that all style-related checks are passing now, @ryan-jacobs, so I think the last steps are to finish the granularity code and then fix all the broken tests. I'm happy to help if you need anything more!

File: src/OaiPmh/Harvester.php
Comment (demiankatz) [inline src/OaiPmh/Harvester.php]: I noticed that you changed "server" to "host" in your last commit... but I thought "server" seemed more accurate. Maybe we should say "OAI server" to be more explicit.

**Review by demiankatz — COMMENTED** [review]
Thanks for the progress on this -- just one very minor comment.

I should also note that I'm going to be out of the office tomorrow, so apologies if I'm not very responsive until next week!

File: src/OaiPmh/Harvester.php
Comment (ryan-jacobs) [reply src/OaiPmh/Harvester.php]: Yeah, I think we have matched intent here. I want to make clear that the time is not based on local VuFind environment. I'll make another edit to make this clear.

File: src/OaiPmh/Harvester.php
Comment (ryan-jacobs) [reply src/OaiPmh/Harvester.php]: Great! I'm glad we talked though that detail. https://github.com/vufind-org/vufindharvest/pull/8/commits/314bf0a1c86a769004f3217c99b27503efcde005 adds the granularity check. I think this means that we also have to maintain some test coverage for differing granularity, but that should not be an issue.

**Comment by ryan-jacobs** [comment]
I think the main updates and style adjustments are now in place. Turning to the tests I committed a set of changes that should get everything passing in https://github.com/vufind-org/vufindharvest/pull/8/commits/abc0af7281cd5bdb8b233a9ffeae74da743b30c5

There were some test removals as part of this, but I think they are well justified given other refactoring that has taken place. I will add some in-line comments that get into that in a moment.

File: tests/integration-tests/src/VuFindTest/OaiPmh/HarvesterFactoryTest.php
Comment (ryan-jacobs) [inline tests/integration-tests/src/VuFindTest/OaiPmh/HarvesterFactoryTest.php]: Explanation: Within the context of HarvesterFactoryTest I don't think it's relevant to be testing any Identify calls or granularity detection anymore. This is because getIdentifyResponse() now allows Identify calls and granularity detection to be done lazily, instead of running as part of Harvester instantiation. I think that justifies the removal of these 3 related tests in HarvesterFactoryTest. Other adjustments in HarvesterTest continue to deliver some coverage for granularity detection.

File: tests/unit-tests/src/VuFindTest/OaiPmh/HarvesterTest.php
Comment (ryan-jacobs) [inline tests/unit-tests/src/VuFindTest/OaiPmh/HarvesterTest.php:269]: Explanation: Similar to the removals in HarvesterFactoryTest we should no longer test granularity detection as a direct property of a new Harvester object.

File: tests/unit-tests/src/VuFindTest/OaiPmh/HarvesterTest.php
Comment (ryan-jacobs) [inline tests/unit-tests/src/VuFindTest/OaiPmh/HarvesterTest.php:411]: Explanation: This method is my attempt to retain some coverage for granularity detection. This simply ensures that we get a YYYY-MM-DD formatted end date into state when day granularity is declared.

File: tests/unit-tests/src/VuFindTest/OaiPmh/HarvesterTest.php
Comment (ryan-jacobs) [inline tests/unit-tests/src/VuFindTest/OaiPmh/HarvesterTest.php:290]: Explanation: This date now comes directly from the mocked Identify responseDate. So asserting a match to that responseDate here should provide coverage for the new end date handling that's central to #7. See also the new testSimpleListRecordsGranularityHandling().

**Comment by demiankatz** [comment]
Thanks, @ryan-jacobs, I think we're very close to the finish line now. Just a couple of comments:

1.) There are a handful of style issues to be addressed; `vendor/bin/phing phpcbf` should automatically fix most of them, but the remainder are long lines that will need to be manually wrapped. Let me know if you need any help with that!

2.) Some of the tests you removed seem to be designed to test features of the VuFindHarvest\OaiPmh\Communicator class which are not covered elsewhere. These tests were clearly not in the most appropriate place, and I'm not sure why we don't have a unit test for the Communicator by itself. I think the answer is probably to create the missing unit test. The only question is whether we try to do that here, or address it in a separate PR. Since this PR doesn't touch the Communicator code at all, opening a separate PR that adds a new unit test to cover the Communicator might be a good way to solve the problem while keeping things less tangled over here. I'm willing to help if necessary! Whatever we decide, it will be a good idea to look at the [Clover coverage report in Jenkins](https://vufind.org/jenkins/job/VuFindHarvest/clover/) after we merge this to make sure we haven't lost any coverage.

**Comment by ryan-jacobs** [comment]
> There are a handful of style issues to be addressed; vendor/bin/phing phpcbf should automatically fix most of them, but the remainder are long lines that will need to be manually wrapped. Let me know if you need any help with that!

Indeed... I missed those. Btw, I wasn't able to see the long-line warnings with `phpcbf` alone, but I was able to surface them with `phpcs-console`. The last couple commits should address them, but let me know if there's any style rules I may be misinterpreting.

> Some of the tests you removed seem to be designed to test features of the VuFindHarvest\OaiPmh\Communicator class which are not covered elsewhere. These tests were clearly not in the most appropriate place, and I'm not sure why we don't have a unit test for the Communicator by itself. I think the answer is probably to create the missing unit test. The only question is whether we try to do that here, or address it in a separate PR

Ahh, I see that now. I thought the client connection bits of these tests were redundant, but I see how they are unique. This is one where I may indeed need some help, but to make sure this is being tracked officially (peace of mind if moving this PR forward) I opened a follow-up with starter approach in PR #10.

**Review by demiankatz — APPROVED** [review]
Thanks again, @ryan-jacobs -- looks like everything is passing now! I'm going to hold off on merging this until #10 is completed, since if we merge #10 first, we'll have no gap in test coverage. Once that's all done, I'll update the changelog and issue a new release!

Also note that I'm on vacation today, but I'll be back to work tomorrow and should be able to finish this up by the end of the week, assuming you're able to finish #10 (and if you need more help there, let me know and I can take over as needed -- there's really not much left to do).

**Review by demiankatz — APPROVED** [review]
I believe we're in good shape now. Thanks again for the thoughtful and thorough work on this, @ryan-jacobs. Merging now, and I'll update the changelog and issue a release shortly!

**Comment by ryan-jacobs** [comment]
Great to see this merged! Thanks @demiankatz for your support and maintainership!

**Pull request merged**
